osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs

```diff
@@ -37,31 +37,33 @@ protected override double StrainValueOf(DifficultyHitObject current)
             var maniaCurrent = (ManiaDifficultyHitObject)current;
             double endTime = maniaCurrent.EndTime;
             int column = maniaCurrent.BaseObject.Column;
+            double closestEndTime = 100;
 
             double holdFactor = 1.0; // Factor to all additional strains in case something else is held
             double holdAddition = 0; // Addition to the current note in case it's a hold and has to be released awkwardly
 
             // Fill up the holdEndTimes array
             for (int i = 0; i < holdEndTimes.Length; ++i)
             {
-                // If there is at least one other overlapping end or note, then we get an addition, buuuuuut...
+                // If there is at least one other overlapping end or note, then we get an addition
                 if (Precision.DefinitelyBigger(holdEndTimes[i], maniaCurrent.StartTime, 1) && Precision.DefinitelyBigger(endTime, holdEndTimes[i], 1))
                     holdAddition = 1.0;
 
-                // ... this addition only is valid if there is _no_ other note with the same ending. Releasing multiple notes at the same time is just as easy as releasing 1
-                if (Precision.AlmostEquals(endTime, holdEndTimes[i], 1))
-                    holdAddition = 0;
-
                 // We give a slight bonus to everything if something is held meanwhile
                 if (Precision.DefinitelyBigger(holdEndTimes[i], endTime, 1))
                     holdFactor = 1.25;
 
+                closestEndTime = Math.Min(closestEndTime, Math.Abs(endTime - holdEndTimes[i]));
+
                 // Decay individual strains
                 individualStrains[i] = applyDecay(individualStrains[i], current.DeltaTime, individual_decay_base);
             }
 
             holdEndTimes[column] = endTime;
 
+            // The hold addition only is valid if there is _no_ other note with the same ending. Releasing multiple notes at the same time is just as easy as releasing 1
+            holdAddition *= 1 / (1 + Math.Exp(0.5 * (24 - closestEndTime)));
```

Comment:
## Code efficiency

As an optimization, this should be done only if `holdAddition` is more than 0.

The `24` looks like a magic number. It actually determines the point at which LN difficulty is nerfed by half. In this case that point is where note endings are 24 ms apart. There should really be a comment regarding this.

## Algorithm concerns

If I understood the code correctly, let's say `endTime` is the red shaded note's end time, and the end times of blue shaded notes are `holdEndTimes[i]`. Numbered below are columns `i`. 

![](https://i.imgur.com/kbLa6tS.png)

Since the red shaded note and the note in column 2 are considered overlapping as per condition on line 49, `holdAddition` would then be `1.0`. Outlined in green are the two notes that have the smallest difference in their end times, which would make `closestEndTime` be 0.  Therefore, `holdAddition` would also be 0.

In short, any synced hold note ends are going to undermine any other hold note ends that have a significant time difference. This extends to the matter that this does not take into account inter-hand patterns. Two note with synced ends on different hands can nerf notes that require significant hand dexterity on single hand.

This doesn't look desirable. The red shaded note is requires non 0 finger dexterity to perform, but this is nerfing it to 0.

As a note for the future in regards to balancing difficulty of LN patterns, I also believe this should be OD dependent.

Reply:
> As an optimization, this should be done only if holdAddition is more than 0.

Done

> The 24 looks like a magic number.

Commented its purpose here.

>In short, any synced hold note ends are going to undermine any other hold note ends that have a significant time difference. 

Totally agree with you on this, but I didn't want to change the original algorithm completely. The main purpose here was to make sure that mirroring the map didn't change the star rating. Taking what hand the note is on should definitely be taken into account in future fixes.

> I also believe this should be OD dependent.

Definitely, but for ranked maps this shouldn't be too much of an issue as the values for OD are not too different.